### PR TITLE
implement hash on basicconstraints

### PR DESCRIPTION
**Pull request by reaperhulk**
If this approach looks good I'll probably do it on more than one extension for the next PR ;)

refs #2578

### CODE REVIEW

**Comment by reaperhulk** [comment]
jenkins, retest this please

File: src/cryptography/x509/extensions.py [hunk 1]
```diff
@@ -342,6 +342,9 @@ def __eq__(self, other):
     def __ne__(self, other):
         return not self == other
 
+    def __hash__(self):
+        return hash(self.ca) ^ hash(self.path_length)
```
Comment (alex) [inline src/cryptography/x509/extensions.py]: So there's this (xoring each of the hashes), or `hash((self.ca, self.path_length))` (xoring a tuple of the items).

https://github.com/python/cpython/blob/master/Objects/tupleobject.c#L337-L359 is the algorithm for tuples, basically a xor + fancy stuff.

opinions?

Comment (reaperhulk) [inline src/cryptography/x509/extensions.py]: Seems like hashing the tuple and letting the underlying implementation xor is reasonable. I'll switch it.

Comment (alex) [inline src/cryptography/x509/extensions.py]: :+1:

On Sun, Dec 27, 2015 at 2:00 PM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/x509/extensions.py
> <https://github.com/pyca/cryptography/pull/2594#discussion_r48457325>:
>
> > @@ -342,6 +342,9 @@ def __eq__(self, other):
> >      def __ne__(self, other):
> >          return not self == other
> >
> > +    def __hash__(self):
> > +        return hash(self.ca) ^ hash(self.path_length)
>
> Seems like hashing the tuple and letting the underlying implementation xor
> is reasonable. I'll switch it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2594/files#r48457325>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: tests/test_x509_ext.py [hunk 2]
```diff
@@ -837,6 +837,11 @@ def test_repr(self):
             "<BasicConstraints(ca=True, path_length=None)>"
         )
 
+    def test_hash(self):
+        na = x509.BasicConstraints(ca=True, path_length=None)
+        na2 = x509.BasicConstraints(ca=True, path_length=None)
+        assert hash(na) == hash(na2)
```
Comment (alex) [inline tests/test_x509_ext.py]: this should have a hash `!=` branch

**Pull request merged**